File: server/api/session.py
"""会话 API：当前项目、统计、开关项目"""
from fastapi import APIRouter, Depends
from pydantic import BaseModel

from ..deps import get_state, require_project
from ..state import AppState

router = APIRouter(tags=['session'])


class OpenRequest(BaseModel):
    name: str


async def _stats(state: AppState) -> dict:
    if state.db is None:
        return {}
    d = await state.db_call(state.db.get_dialogue_count)
    u = await state.db_call(state.db.get_ui_text_count)
    n = await state.db_call(state.db.get_char_dict_count)
    return {'dialogue': d, 'ui': u, 'names': n}


@router.get('/session')
async def get_session(state: AppState = Depends(get_state)):
    return {
        'current_project': state.current_project,
        'has_translator': state.translator is not None,
        'stats': await _stats(state),
        'active_jobs': state.app_db.list_jobs(active_only=True),
        'interrupted_jobs': state.interrupted_count,
    }


@router.post('/session/open')
async def open_project(req: OpenRequest, state: AppState = Depends(get_state)):
    await state.open_project(req.name)
    return {
        'current_project': state.current_project,
        'has_translator': state.translator is not None,
        'stats': await _stats(state),
    }


@router.post('/session/close')
async def close_project(state: AppState = Depends(require_project)):
    await state.close_project()
    return {'current_project': ''}

File: web/src/stores/session.ts
/** 会话 store：当前项目 + 统计 + 活跃任务 */
import { defineStore } from 'pinia'
import { api } from '../api/client'

export interface CountStat {
  total: number
  translated: number
  untranslated: number
}

export interface SessionStats {
  dialogue?: CountStat
  ui?: CountStat
  names?: CountStat
}

export const useSessionStore = defineStore('session', {
  state: () => ({
    currentProject: '' as string,
    hasTranslator: false,
    stats: {} as SessionStats,
    interruptedJobs: 0,
    loaded: false,
  }),
  getters: {
    progressText(state): string {
      // 只显示有内容的类别；都没有则空
      const parts: string[] = []
      const d = state.stats.dialogue
      const u = state.stats.ui
      const n = state.stats.names
      if (d?.total) parts.push(`对话 ${d.translated}/${d.total}`)
      if (u?.total) parts.push(`字符串 ${u.translated}/${u.total}`)
      if (n?.total) parts.push(`人物 ${n.translated}/${n.total}`)
      return parts.length ? parts.join(' · ') : ''
    },
  },
  actions: {
    async refresh() {
      const data = await api.get<{
        current_project: string
        has_translator: boolean
        stats: SessionStats
        interrupted_jobs?: number
      }>('/api/session')
      this.currentProject = data.current_project
      this.hasTranslator = data.has_translator
      this.stats = data.stats
      this.interruptedJobs = data.interrupted_jobs ?? 0
      this.loaded = true
    },
    async open(name: string) {
      const data = await api.post<{
        current_project: string
        has_translator: boolean
        stats: SessionStats
      }>('/api/session/open', { name })
      this.currentProject = data.current_project
      this.hasTranslator = data.has_translator
      this.stats = data.stats
    },
    async close() {
      await api.post('/api/session/close')
      this.currentProject = ''
      this.stats = {}
    },
  },
})
